File: src/main/scala/com/microsoft/azure/cosmosdb/kafka/connect/sink/CosmosDBSinkTask.scala
package com.microsoft.azure.cosmosdb.kafka.connect.sink

import java.util

import scala.collection.mutable.HashMap
import com.microsoft.azure.cosmosdb.kafka.connect.config.{ConnectorConfig, CosmosDBConfig, CosmosDBConfigConstants}
import com.microsoft.azure.cosmosdb.kafka.connect.{CosmosDBClientSettings, CosmosDBProviderImpl, CosmosDBProvider}
import com.microsoft.azure.cosmosdb.kafka.connect.processor._
import com.microsoft.azure.cosmosdb.rx.AsyncDocumentClient
import com.microsoft.azure.cosmosdb.{ConnectionPolicy, ConsistencyLevel}
import com.typesafe.scalalogging.LazyLogging
import org.apache.kafka.clients.consumer.OffsetAndMetadata
import org.apache.kafka.common.TopicPartition
import org.apache.kafka.connect.errors.ConnectException
import org.apache.kafka.connect.sink.{SinkRecord, SinkTask}
import scala.collection.JavaConverters._
import scala.util.{Failure, Success, Try}

class CosmosDBSinkTask extends SinkTask with LazyLogging {

  private var writer: Option[CosmosDBWriter] = None

  private var client: AsyncDocumentClient = null
  private var database: String = ""
  private var taskConfig: Option[CosmosDBConfig] = None
  private var topicNames: Array[String] = null
  private var postProcessors  = List.empty[PostProcessor]
  val collectionTopicMap: HashMap[String, String] = HashMap.empty[String, String] // Public to allow for testing
  val cosmosDBProvider: CosmosDBProvider = CosmosDBProviderImpl

  override def start(props: util.Map[String, String]): Unit = {
    logger.info("Starting CosmosDBSinkTask")

    var config: util.Map[String, String] = null
    if (context != null) {
        config = if (context.configs().isEmpty) props else context.configs()
    }
    else {
        config = props
    }

    // Get Configuration for this Task
    taskConfig = Try(CosmosDBConfig(ConnectorConfig.sinkConfigDef, config)) match {
        case Failure(f) => throw new ConnectException("Couldn't start CosmosDBSink due to configuration error.", f)
        case Success(s) => Some(s)
    }

    // Add configured Post-Processors
    val processorClassNames = taskConfig.get.getString(CosmosDBConfigConstants.SINK_POST_PROCESSOR)
    postProcessors = PostProcessor.createPostProcessorList(processorClassNames, taskConfig.get)

    // Get CosmosDB Connection
    val endpoint: String = taskConfig.get.getString(CosmosDBConfigConstants.CONNECTION_ENDPOINT_CONFIG)
    val masterKey: String = taskConfig.get.getPassword(CosmosDBConfigConstants.CONNECTION_MASTERKEY_CONFIG).value()
    database = taskConfig.get.getString(CosmosDBConfigConstants.DATABASE_CONFIG)

    // Populate collection topic map
    // TODO: add support for many to many mapping, this only assumes each topic writes to one collection and multiple topics can write to the same collection
    val collectionTopicMapString = taskConfig.get.getString(CosmosDBConfigConstants.COLLECTION_TOPIC_MAP_CONFIG)
    if(collectionTopicMapString.contains("#")) { // There is at least one pair
        collectionTopicMapString.split(",").map(_.trim).foreach(
            m => {
                val map = m.split("#").map(_.trim)
                collectionTopicMap.put(map(1), map(0)) // topic, collection
            })
    }

    // If there are topics with no mapping, add them to the map with topic name as collection name
    topicNames = taskConfig.get.getString(CosmosDBConfigConstants.TOPIC_CONFIG).split(",").map(_.trim)
    topicNames.foreach(
        t => {
            if (!collectionTopicMap.contains(t)) {
                collectionTopicMap.put(t, t) // topic, collection
            }
        })

    val clientSettings = CosmosDBClientSettings(
        endpoint,
        masterKey,
        database,
        null,   // Don't pass a collection because our client is potentially for multiple collections
        ConnectionPolicy.GetDefault(),
        ConsistencyLevel.Session
    )
    client = Try(cosmosDBProvider.getClient(clientSettings)) match {
        case Success(conn) =>
            logger.info("Connection to CosmosDB established.")
            conn
        case Failure(f) => throw new ConnectException(s"Couldn't connect to CosmosDB.", f)
    }

    // Set up Writer
    val setting = new CosmosDBSinkSettings(endpoint, masterKey, database, collectionTopicMap)
    writer = Option(new CosmosDBWriter(setting, cosmosDBProvider))
  }


  override def put(records: util.Collection[SinkRecord]): Unit = {
    val seq = records.asScala.toList
    logger.info(s"Sending ${seq.length} records to writer to be written")

    // Execute PostProcessing
    val postProcessed = seq.map(sr => applyPostProcessing(sr))

    // Currently only built for messages with JSON payload without schema
    writer.foreach(w => w.write(postProcessed))
  }

  override def stop(): Unit = {
    logger.info("Stopping CosmosDBSinkTask")
  }

  override def flush(map: util.Map[TopicPartition, OffsetAndMetadata]): Unit = {}

  override def version(): String = getClass.getPackage.getImplementationVersion

  private def applyPostProcessing(sinkRecord: SinkRecord): SinkRecord =
    postProcessors.foldLeft(sinkRecord)((r, p) => {
      //println(p.getClass.toString)
      p.runPostProcess(r)
    })

}

File: src/main/java/com/microsoft/azure/cosmosdb/kafka/connect/sink/CosmosDBSinkTask.java
package com.microsoft.azure.cosmosdb.kafka.connect.sink;

import com.microsoft.azure.cosmosdb.kafka.connect.SettingDefaults;
import com.azure.cosmos.CosmosClient;
import com.azure.cosmos.CosmosClientBuilder;
import com.azure.cosmos.CosmosContainer;
import com.azure.cosmos.implementation.BadRequestException;
import org.apache.commons.collections4.CollectionUtils;
import org.apache.kafka.connect.sink.SinkRecord;
import org.apache.kafka.connect.sink.SinkTask;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.apache.kafka.connect.data.Struct;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.util.Collection;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * Implements the Kafka Task for the CosmosDB Sink Connector
 */
public class CosmosDBSinkTask extends SinkTask {
    private static final Logger logger = LoggerFactory.getLogger(CosmosDBSinkTask.class);
    private CosmosClient client = null;
    private SinkSettings settings = null;
    ObjectMapper mapper = new ObjectMapper();

    @Override
    public String version() {
        return this.getClass().getPackage().getImplementationVersion();
    }

    @Override
    public void start(Map<String, String> map) {
        logger.trace("Sink task started.");
        this.settings = new SinkSettings();
        this.settings.populate(map);

        this.client = new CosmosClientBuilder()
                .endpoint(settings.getEndpoint())
                .key(settings.getKey())
                .userAgentSuffix(SettingDefaults.COSMOS_CLIENT_USER_AGENT_SUFFIX+version())
                .buildClient();

        client.createDatabaseIfNotExists(settings.getDatabaseName());
    }

    @Override
    public void put(Collection<SinkRecord> records) {
        if (CollectionUtils.isEmpty(records)) {
            logger.info("No records to be written");
            return;
        }

        logger.info("Sending {} records to be written", records.size());

        Map<String, List<SinkRecord>> recordsByContainer = records.stream()
                // Find target collection for each record
                .collect(Collectors.groupingBy(record -> settings.getTopicContainerMap()
                        .getContainerForTopic(record.topic()).orElseThrow(() -> new IllegalStateException(
                                String.format("No container defined for topic %s .", record.topic())))));

        for (Map.Entry<String, List<SinkRecord>> entry : recordsByContainer.entrySet()) {
            String containerName = entry.getKey();
            CosmosContainer container = client.getDatabase(settings.getDatabaseName()).getContainer(containerName);
            for (SinkRecord record : entry.getValue()) {
                logger.debug("Writing record, value type: {}", record.value().getClass().getName());
                logger.debug("Key Schema: {}", record.keySchema());
                logger.debug("Value schema: {}", record.valueSchema());
                logger.debug("Value.toString(): {}", record.value() != null ? record.value().toString() : "<null>");

                Object recordValue;
                if (record.value() instanceof Struct) {
                    Map<String, Object> jsonMap = StructToJsonMap.toJsonMap((Struct) record.value());
                    recordValue = mapper.convertValue(jsonMap, JsonNode.class);
                } else {
                    recordValue = record.value();
                }

                try {
                    if (Boolean.TRUE.equals(this.settings.getUseUpsert())) {
                        container.upsertItem(recordValue);
                    } else {
                        container.createItem(recordValue);
                    }
                } catch (BadRequestException bre) {
                    throw new CosmosDBWriteException(record, bre);
                }
            }
        }
    }

    @Override
    public void stop() {
        logger.trace("Stopping CosmosDB sink task");

        if (client != null) {
            client.close();
            client = null;
        }
        
        settings = null;

    }
}
